src/core/public/application/application_service.tsx

```diff
@@ -151,6 +151,10 @@ export class ApplicationService {
 
     const wrapMount = (plugin: PluginOpaqueId, app: App<any>): AppMount => {
       return async (params) => {
+        const currentAppId = this.currentAppId$.value;
+        if (currentAppId && currentAppId !== app.id) {
+          this.appInternalStates.delete(currentAppId);
+        }
```

Comment:
Improve consistency by performing the 'on entering new app' logic at a single place now (the internal state was not cleaned when changing app via history, this was a bug)